src/calculate/pass_4/rah.rs

```diff
@@ -0,0 +1,199 @@
+use crate::calculate::item::Object;
+use crate::calculate::pass_3::Cache;
+use crate::calculate::Ship;
+use crate::info::Info;
+
+const RAH_TYPE_ID: i32 = 4403;
+const RESISTANCE_SHIFT_AMOUNT_ID: i32 = 1849;
+const RESONANCE_IDS: [i32; 4] = [267, 268, 269, 270]; // Order: Em, Explo, Kin, Therm
+
+pub fn adapt_rah_to_damage_pattern(info: &impl Info, ship: &mut Ship) {
+    let rah_index = if let Some(i) = ship
+        .items
+        .iter_mut()
+        .position(|item| item.type_id == RAH_TYPE_ID)
+    {
+        i
+    } else {
+        return;
+    };
+
+    let effective_resonance_ids = [
+        info.attribute_name_to_id("armorEmDamageEffectiveResonance"),
+        info.attribute_name_to_id("armorExplosiveDamageEffectiveResonance"),
+        info.attribute_name_to_id("armorKineticDamageEffectiveResonance"),
+        info.attribute_name_to_id("armorThermalDamageEffectiveResonance"),
+    ];
+
+    let rah_affected_attributes = rah_affected_attributes(ship, rah_index);
+    let shift_amount = ship.items[rah_index]
+        .attributes
+        .get(&RESISTANCE_SHIFT_AMOUNT_ID)
+        .and_then(|a| a.value)
+        .unwrap()
+        / 100.0;
+
+    let mut cache = Cache::default();
+    let mut cycle_list: Vec<[f64; 4]> = Vec::new();
+    let mut cycle_start = None;
+
+    // The RAH changes its resistances every cycle based on the damaged received on that cycle.
```

Comment:
```suggestion
    // The RAH changes its resistances every cycle based on the effective damaged received on that cycle.
```

That is the key point why simulation is needed. It is not based on the damage, but on the effective damage. This means a change in resistance can change the damage type that does the most damage :)